### PR TITLE
Fix  recursive spinlock in k_msgq_get

### LINKED ISSUE #45267: kernel: Recursive spinlock in k_msgq_get() in the context of a k_work_poll handler
**Describe the bug**
Recursive spinlock in `k_msgq_get()`, when called from a k_work_poll handler function, when event is configured for notification on message availability in the MsgQ.

This situation appears only if the thread "putting" the message in the MsgQ is preemptive and if the workqueue assigned for the process of the k_work_poll handler has a higher priority.

**Edit: I described the issue for the MsgQ only but actually this issue occurs with all "transferable" kernel objects which support POLLING** :
- MsgQ: `k_msgq_put()` / `k_msgq_get()` through `K_POLL_TYPE_MSGQ_DATA_AVAILABLE`
- Fifo: `k_fifo_put()` / `k_fifo_get()` through `K_POLL_TYPE_FIFO_DATA_AVAILABLE`
- Semaphores: `k_sem_give()` / `k_sem_take()` through `K_POLL_TYPE_SEM_AVAILABLE`

**To Reproduce**
Please find here this Zephyr project [zephyr-qemu-dev](https://github.com/lucasdietrich/zephyr-qemu-dev/tree/k_work_poll_submit_test) ( [msgq: main.c](https://github.com/lucasdietrich/zephyr-qemu-dev/blob/k_work_poll_submit_test/src/main.c) ) which can be run in qemu to reproduce this issue. (Same issue with FIFO [fifo: main.c](https://github.com/lucasdietrich/zephyr-qemu-dev/blob/k_work_poll_submit_test_fifo/src/main.c))

Basically:
- Enable `CONFIG_POLL` and `CONFIG_ASSERT`
- Create a **preemptive** thread with a certain priority (e.g. `K_PRIO_PREEMPT(12)`)
- Create a workqueue with a **higher** priority (e.g. `K_PRIO_COOP(1)`)
- Configure a k_work_poll item to be notified for any new message in a MsgQ (`K_POLL_TYPE_MSGQ_DATA_AVAILABLE`), which will queue the work item to the configured workqueue.
- Configure the k_work_poll handler function to do a `k_msg_get` on the MsgQ when called
- From the thread, "put" a message to the MsgQ using `k_msgq_put()`

When running this application, the execution will fail with an "assertion error" (if CONFIG_ASSERT is enabled):

As far as I remember, when CONFIG_ASSERT is disabled the k_work_poll is not resubmitted correctly (to confirm).

**Expected behavior**
No assertion error.

**Impact**
It's then not possible to handle msgQ messages directly from the k_work_poll handler and the k_work_poll item cannot be resubmitted from the same handler using `k_work_pull_submit(_to_queue)`. A workarround would be to create a dedicated thread which waits on MsgQ messages instead of a k_work_poll.

**Logs and console output**
The assertion error occurs [here](https://github.com/zephyrproject-rtos/zephyr/blob/83c79d10510e28e8706581168d257c1b556732d2/kernel/msg_q.c#L214)
```
*** Booting Zephyr OS build zephyr-v2.7.2  ***
ASSERTION FAIL [z_spin_lock_valid(l)] @ WEST_TOPDIR/zephyr/include/spinlock.h:142
        Recursive spinlock 0x10a130
```

**Environment :**
 - Linux
 - Zephyr 2.7.2, 3.0.0 and current main (83c79d10510e28e8706581168d257c1b556732d2)
 - All platforms

**Diagnostic**
I debugged this and noticed that a `k_yield()` is called from `k_work_submit_to_queue()` in the context of the MsgQ lock which may cause a thread switch before unlocking the `msgq->lock` (see the call stack in the screenshot).

https://github.com/zephyrproject-rtos/zephyr/blob/83c79d10510e28e8706581168d257c1b556732d2/kernel/work.c#L376-L378

issue_z_impl_k_msgq_put_2

If the program follows this particular path, because of this `k_yield()`, the work item handler will be executed while the `msgq->lock` stays locked by the thread calling `k_msgq_put()`. Then the call to `k_msgq_get()` in the work handler will cause a "Recursive spinlock" assertion error.

Disabling the `k_yield()` solves the issue : https://github.com/lucasdietrich/zephyr/commit/50923a3f7b1dbd04c52b54aaa5ac2d97e0253bea

**Test**
I already wrote this test routine to cover this situation (for which I could propose a pull request) : https://github.com/lucasdietrich/zephyr/commit/c0a7deb7b268a794e811bc976c3b2eb7322484c0, run the test with `./scripts/twister -i -v -p qemu_x86 -T tests/kernel/workq/work_queue`

**Solutions**
What would be good solutions for this ? This `k_yield()` should definitely not be called in this context, but how to notify the `k_work_submit_to_queue()` to not yield in this case ?

A simple and naive solution is, but which I think is terrible :
https://github.com/lucasdietrich/zephyr/commit/9c49c67d5577b86d213e62b666360ea587395134#diff-06e616b4678c2bb4ee4b383f66a36821aa1940dc7bf20b8a3f68b1573bb02875

Thank you for your help 😄

### PR DESCRIPTION
**Pull request by lucasdietrich**
Proposition for [solution 2](https://github.com/zephyrproject-rtos/zephyr/issues/45267#issuecomment-1116361810) to fix issue #45267 + test

Fixes #45267

### CODE REVIEW

File: kernel/work.c [hunk 1]
```diff
@@ -355,19 +355,38 @@ static int submit_to_queue_locked(struct k_work *work,
...
+	return ret;
+}
+
+int k_work_submit_to_queue(struct k_work_q *queue,
+			    struct k_work *work)
+{
+	SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_work, submit_to_queue, queue, work);
+
+	int ret = z_work_submit_to_queue(queue, work);
+
```
Comment (andyross) [inline kernel/work.c:389]: FWIW: I think this patch crossed with one of mine that just went in hitting exactly these lines (it replaces the yield with a proper reschedule), see commit https://github.com/zephyrproject-rtos/zephyr/commit/8d94967ec4773d9af67cb70167fe765085f3f737  Presumably it needs a rebase?

File: tests/kernel/workq/work_queue/src/main.c [hunk 2]
```diff
@@ -701,12 +725,79 @@ static void test_triggered_wait_expired(void)
...
+}
+
+static void triggered_from_msgq_work_handler(struct k_work *work)
+{
+	msg_t msg;
+
+	zassert_true(k_msgq_get(&triggered_from_msgq_test.msgq, &msg, K_NO_WAIT) == 0, NULL);
+}
+
+static void test_triggered_from_msgq_init(void)
```
Comment (andyross) [inline tests/kernel/workq/work_queue/src/main.c:747]: Pedantry, but this is one of those situations where you're testing for a regression that is going to seem really obscure to someone reading this code in the future.  Might be a good idea to explain it here, or just link to the github issue.

**Review by andyross — DISMISSED** [review]
This definitely looks correct to me.  Giving it a +1 here for posterity, even though I'm sure you'll have to wipe that out with a rebase and I'll need to be bugged to come look at it again.

**Review by andyross — DISMISSED** [review]
This looks good to me.  Note that:

> This solves https://github.com/zephyrproject-rtos/zephyr/issues/45267

If you rephrase that to "Fixes #45267" then the merge will automatically close the bug.

**Review by andyross — APPROVED** [review]
Refresh +1, looks good to me.

**Review by dcpleung — APPROVED** [review]

**Pull request merged**